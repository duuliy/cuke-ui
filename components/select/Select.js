import React, { PureComponent, createRef } from "react";
import { createPortal } from "react-dom";
import PropTypes from "prop-types";
import cls from "classnames";
import Input from "../input";
import { DownIcon, LoadingIcon } from "../icon";
import { debounce } from "../utils";
import scrollIntoViewIfNeeded from "scroll-into-view-if-needed";
import Empty from "../empty";

const sizes = {
  default: "default",
  small: "small",
  large: "large"
};

export default class Select extends PureComponent {
  state = {
    selectedValue: this.props.defaultValue || this.props.value || "",
    visible: this.props.visible || null,
    left: 0,
    top: 0,
    width: 0
  };
  static defaultProps = {
    prefixCls: "cuke-select",
    onPanelVisibleChange: () => {},
    onChange: () => {},
    getPopupContainer: () => document.body,
    position: "bottom",
    disabled: false,
    allowClear: false,
    notFoundContent: <Empty height={120} />,
    labelInValue: false,
    loading: false
  };
  static propTypes = {
    prefixCls: PropTypes.string.isRequired,
    onPanelVisibleChange: PropTypes.func,
    getPopupContainer: PropTypes.func,
    onChange: PropTypes.func,
    disabled: PropTypes.bool,
    allowClear: PropTypes.bool,
    labelInValue: PropTypes.bool,
    size: PropTypes.oneOf(Object.values(sizes)),
    overlay: PropTypes.oneOfType([
      PropTypes.element,
      PropTypes.string,
      PropTypes.object
    ]),
    popupContainerClassName: PropTypes.string
  };
  constructor(props) {
    super(props);
    this.timeOutId = null;
    this.toggleContainer = createRef();
    this.triggerWrapper = createRef();
    this.wrapper = createRef();
  }

  static getDerivedStateFromProps(props) {
    if (Reflect.has(props, "value")) {
      return {
        selectedValue: props.value
      };
    }
    return null;
  }

  get selectedValue() {
    const { selectedValue } = this.state;
    const { labelInValue } = this.props;
    if (labelInValue) {
      // children 如果有空格 会被解析成数组 有可能是 [10,"条数据"] 所以需要 join 一下
      return Array.prototype.join.call(
        (
          this.selectOptions.find(
            ({ key }) =>
              key === ((selectedValue && selectedValue.key) || selectedValue)
          ) || {}
        ).value,
        ""
      );
    }
    return selectedValue;
  }

  get selectOptions() {
    return this.props.children.map(({ props }) => {
      return {
        key: props.value,
        value: props.children
      };
    });
  }

  onChange = (value, label) => {
    const { labelInValue } = this.props;
    this.setState({ selectedValue: value, visible: false });
    this.props.onPanelVisibleChange(false);
    if (labelInValue) {
      this.props.onChange({
        key: value,
        label
      });
    } else {
      this.props.onChange(value);
    }
  };
  onClickHandler = () => {
    const visible = !this.state.visible;
    this.setState({
      visible
    });
    this.props.onPanelVisibleChange(visible);
    if (visible) {
      this.setWrapperBounding(() => {
        scrollIntoViewIfNeeded(this.wrapper.current, {
          scrollMode: "if-needed",
          behavior: "smooth",
          block: "nearest",
          inline: "nearest"
        });
      });
    }
  };
  onClickOutsideHandler = e => {
    e.stopPropagation();
    if (
      this.state.visible &&
      !this.props.disabled &&
      !this.toggleContainer.current.contains(e.target) &&
      !e.target.classList.contains(`${this.props.prefixCls}-option-disabled`)
    ) {
      this.setState({ visible: false });
      this.props.onPanelVisibleChange(false);
    }
  };
  getWrapperBounding = () => {
    const {
      width,
      height,
      top,
      left
    } = this.triggerWrapper.current.getBoundingClientRect();
    const {
      height: wrapperHeight
    } = this.wrapper.current.getBoundingClientRect();

    const { scrollX, scrollY } = window;

    const positions = {
      top: {
        top: top + scrollY - wrapperHeight - 10,
        left: left + scrollX,
        width
      },
      bottom: {
        top: top + height + scrollY,
        left: left + scrollX,
        width
      }
    };
    return positions[this.props.position];
  };

  setWrapperBounding(cb = () => {}) {
    const { left, top, width } = this.getWrapperBounding();
    this.setState({ left, top, width }, cb);
  }

  onResizeHandler = debounce(() => {
    this.setWrapperBounding();
  }, 500);

  onClear = () => {
    this.setState({ selectedValue: "", visible: false });
    this.props.onChange("");
  };

  render() {
    const { visible, left, top, width } = this.state;
    const {
      prefixCls,
      className,
      disabled,
      placeholder,
      children,
      getPopupContainer,
      size,
      style,
      allowClear,
      notFoundContent,
      loading,
      popupContainerClassName,
      labelInValue, //eslint-disable-line
      onPanelVisibleChange, //eslint-disable-line
      ...attr
    } = this.props;

    return (
      <div
        className={cls(`${prefixCls}`, className)}
        {...attr}
        style={style}
        ref={this.toggleContainer}
      >
        <div
          className={cls(`${prefixCls}-inner`, {
            [`${prefixCls}-active`]: visible
          })}
          ref={this.triggerWrapper}
        >
          <Input
            disabled={disabled}
            readonly
            placeholder={placeholder}
            className={cls(`${prefixCls}-input`)}
            value={this.selectedValue}
            onClick={this.onClickHandler}
            size={size}
            style={{
              width: style && style.width
            }}
            suffix={
              loading ? (
                <LoadingIcon className={`${prefixCls}-loading`} />
              ) : (
                <DownIcon className={`${prefixCls}-arrow`} />
              )
            }
            allowClear={allowClear}
            onClear={this.onClear}
          />
        </div>
        {createPortal(
          <div
            className={cls(`${prefixCls}-content`, popupContainerClassName, {
              [`${prefixCls}-open`]: visible,
              [`${prefixCls}-close`]: !visible,
              ["cuke-ui-no-animate"]: visible === null
            })}
            ref={this.wrapper}
            style={{
              width,
              left,
              top
            }}
          >
            {children
              ? React.Children.map(children, (element, index) => {
                  return React.cloneElement(element, {
                    key: index,
                    selectedValue: this.selectedValue,
                    onChange: this.onChange
                  });
                })
              : notFoundContent}
          </div>,
          getPopupContainer()
        )}
      </div>
    );
  }
  componentWillUnmount() {
    window.removeEventListener("click", this.onClickOutsideHandler, false);
    window.removeEventListener("resize", this.onResizeHandler);
  }
  componentDidMount() {
    window.addEventListener("click", this.onClickOutsideHandler, false);
    window.addEventListener("resize", this.onResizeHandler);
    if (this.props.visible) {
      this.setWrapperBounding();
    }
  }
}
