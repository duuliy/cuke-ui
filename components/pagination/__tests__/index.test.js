import Select from "../../select";
import React from "react";
import { render, shallow } from "enzyme";
import assert from "power-assert";
import toJson from "enzyme-to-json";
import Pagination from "../index";
import Button from "../../button";
import { ArrowLeftIcon, ArrowRightIcon, SuccessIcon } from "../../icon";
import NumberInput from "../../number-input";

describe("<Pagination/>", () => {
  it("should render Pagination", () => {
    const wrapper = render(
      <div>
        <Pagination current={1} total={50} />
        <Pagination
          current={1}
          total={50}
          locale={{ prevText: "后退", nextText: "前进" }}
        />
        <Pagination current={1} total={50} showQuickJumper />
        <Pagination current={1} total={50} showSizeChanger />
      </div>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should render simple Pagination", () => {
    const wrapper = render(
      <Pagination current={1} total={10} simple separator="|" />
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should render sizes", () => {
    const wrapper = render(
      <div>
        <Pagination current={1} total={50} size="small" showSizeChanger />
        <Pagination current={1} total={50} size="default" showSizeChanger />
        <Pagination current={1} total={50} size="large" showSizeChanger />
      </div>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should render custom separator", () => {
    const wrapper = shallow(
      <Pagination current={1} total={10} simple separator={<SuccessIcon />} />
    );
    expect(wrapper.find(SuccessIcon)).toHaveLength(1);
  });

  it("should find cuke-pagination classnames", () => {
    const wrapper = shallow(<Pagination current={1} total={10} />);
    assert(wrapper.find(".cuke-pagination").length === 1);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should find cuke-pagination-simple classnames when is simple mode", () => {
    const wrapper = shallow(<Pagination current={1} total={10} simple />);
    assert(wrapper.find(".cuke-pagination-simple").length === 1);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should find custom locale text", () => {
    const wrapper = render(
      <Pagination
        current={1}
        total={10}
        locale={{ prevText: "后退", nextText: "前进" }}
      />
    );
    expect(wrapper.text()).toContain("后退");
    expect(wrapper.text()).toContain("前进");
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should find default locale text", () => {
    const wrapper = shallow(<Pagination current={1} total={10} />);
    assert(wrapper.find(ArrowLeftIcon).length === 1);
    assert(wrapper.find(ArrowRightIcon).length === 1);
  });

  it("should emit onChange events", () => {
    const onChange = jest.fn();
    const wrapper = shallow(
      <Pagination onChange={onChange} current={2} total={50} />
    );

    wrapper.find(".cuke-pagination-prev").simulate("click");
    expect(onChange).toHaveBeenCalled();
    expect(wrapper.state().current).toBe(1);

    wrapper.find(".cuke-pagination-next").simulate("click");
    expect(onChange).toHaveBeenCalled();
    expect(wrapper.state().current).toBe(2);
  });

  it("should emit onChange events with simple mode", () => {
    const onChange = jest.fn();
    const wrapper = shallow(
      <Pagination simple onChange={onChange} current={1} total={10} />
    );

    wrapper
      .find(Button)
      .at(1)
      .simulate("click");
    expect(onChange).toHaveBeenCalled();
  });

  it("should emit on click events with simple", () => {
    const onChange = jest.fn();
    const wrapper = shallow(
      <Pagination simple onChange={onChange} current={1} total={10} />
    );
    wrapper
      .find(Button)
      .at(1)
      .simulate("click");
    expect(wrapper.state().current).toBe(2);
  });

  it("should emit onChange events", () => {
    const onChange = jest.fn();
    const wrapper = shallow(
      <Pagination onChange={onChange} current={2} total={50} />
    );

    wrapper.find(".cuke-pagination-prev").simulate("click");
    expect(onChange).toHaveBeenCalled();
    expect(wrapper.state().current).toBe(1);

    wrapper.find(".cuke-pagination-next").simulate("click");
    expect(onChange).toHaveBeenCalled();
    expect(wrapper.state().current).toBe(2);
  });

  it("should render 10 page button when pageSize is 5", () => {
    const wrapper = shallow(<Pagination pageSize={5} total={50} />);
    const prevAndNext = 2;
    expect(wrapper.find(".cuke-pagination-item")).toHaveLength(
      10 + prevAndNext
    );
  });
  it("should render 10 page button when pageSize is 5", () => {
    const wrapper = shallow(<Pagination pageSize={5} total={49} />);
    const prevAndNext = 2;
    expect(wrapper.find(".cuke-pagination-item")).toHaveLength(
      10 + prevAndNext
    );
  });
  it("should render show size changer selector", () => {
    const wrapper = shallow(
      <Pagination pageSize={5} total={49} showSizeChanger />
    );
    expect(wrapper.find(Select)).toHaveLength(1);
  });
  it("should render custom total", () => {
    const wrapper = shallow(
      <Pagination
        pageSize={5}
        total={49}
        showTotal={total => `共${total}条数据`}
      />
    );
    expect(wrapper.text()).toContain("共49条数据");
  });
  it("should trigger onChange", () => {
    const onChange = jest.fn();
    const wrapper = shallow(
      <Pagination pageSize={5} total={49} onChange={onChange} />
    );
    wrapper
      .find(".cuke-pagination-item")
      .at(2)
      .simulate("click");
    expect(onChange).toHaveBeenCalled();
  });
  it("should find quick jumper", () => {
    const wrapper = shallow(
      <Pagination pageSize={5} total={49} showQuickJumper />
    );
    expect(wrapper.find(NumberInput)).toHaveLength(1);
  });
});
