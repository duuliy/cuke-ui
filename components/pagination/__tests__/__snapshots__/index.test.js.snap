// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination/> should find cuke-pagination classnames 1`] = `
<ul
  className="cuke-pagination cuke-pagination-default"
  current={1}
>
  <li
    className="cuke-pagination-show-total"
  />
  <li
    className="cuke-pagination-item cuke-pagination-prev cuke-pagination-item-disabled"
  >
    <ArrowLeftIcon />
  </li>
  <li
    className="cuke-pagination-item cuke-pagination-item-selected"
    key="0"
    onClick={[Function]}
  >
    1
  </li>
  <li
    className="cuke-pagination-item cuke-pagination-next cuke-pagination-item-disabled"
  >
    <ArrowRightIcon />
  </li>
</ul>
`;

exports[`<Pagination/> should find cuke-pagination-simple classnames when is simple mode 1`] = `
<div
  className="cuke-pagination cuke-pagination-simple"
  current={1}
>
  <Button
    block={false}
    circle={false}
    dashed={false}
    disabled={true}
    hollow={false}
    href=""
    htmlType="button"
    loading={false}
    onClick={[Function]}
    plain={false}
    prefixCls="cuke-button"
    size="small"
    type="default"
  >
    <ArrowLeftIcon />
  </Button>
  <span
    className="cuke-pagination-simple-pages"
  >
    <span
      className="cuke-pagination-simple-page-index"
    >
      1
    </span>
    <span
      className="cuke-pagination-simple-page-separator"
    >
      /
    </span>
    <span>
      10
    </span>
  </span>
  <Button
    block={false}
    circle={false}
    dashed={false}
    disabled={false}
    hollow={false}
    href=""
    htmlType="button"
    loading={false}
    onClick={[Function]}
    plain={false}
    prefixCls="cuke-button"
    size="small"
    type="primary"
  >
    <ArrowRightIcon />
  </Button>
</div>
`;

exports[`<Pagination/> should find custom locale text 1`] = `
<ul
  class="cuke-pagination cuke-pagination-default"
  current="1"
>
  <li
    class="cuke-pagination-show-total"
  />
  <li
    class="cuke-pagination-item cuke-pagination-prev cuke-pagination-item-disabled"
  >
    后退
  </li>
  <li
    class="cuke-pagination-item cuke-pagination-item-selected"
  >
    1
  </li>
  <li
    class="cuke-pagination-item cuke-pagination-next cuke-pagination-item-disabled"
  >
    前进
  </li>
</ul>
`;

exports[`<Pagination/> should render Pagination 1`] = `
<div>
  <ul
    class="cuke-pagination cuke-pagination-default"
    current="1"
  >
    <li
      class="cuke-pagination-show-total"
    />
    <li
      class="cuke-pagination-item cuke-pagination-prev cuke-pagination-item-disabled"
    >
      <svg
        fill="none"
        height="1em"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1em"
      >
        <polyline
          points="15 18 9 12 15 6"
        />
      </svg>
    </li>
    <li
      class="cuke-pagination-item cuke-pagination-item-selected"
    >
      1
    </li>
    <li
      class="cuke-pagination-item"
    >
      2
    </li>
    <li
      class="cuke-pagination-item"
    >
      3
    </li>
    <li
      class="cuke-pagination-item"
    >
      4
    </li>
    <li
      class="cuke-pagination-item"
    >
      5
    </li>
    <li
      class="cuke-pagination-item cuke-pagination-next"
    >
      <svg
        fill="none"
        height="1em"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1em"
      >
        <polyline
          points="9 18 15 12 9 6"
        />
      </svg>
    </li>
  </ul>
  <ul
    class="cuke-pagination cuke-pagination-default"
    current="1"
  >
    <li
      class="cuke-pagination-show-total"
    />
    <li
      class="cuke-pagination-item cuke-pagination-prev cuke-pagination-item-disabled"
    >
      后退
    </li>
    <li
      class="cuke-pagination-item cuke-pagination-item-selected"
    >
      1
    </li>
    <li
      class="cuke-pagination-item"
    >
      2
    </li>
    <li
      class="cuke-pagination-item"
    >
      3
    </li>
    <li
      class="cuke-pagination-item"
    >
      4
    </li>
    <li
      class="cuke-pagination-item"
    >
      5
    </li>
    <li
      class="cuke-pagination-item cuke-pagination-next"
    >
      前进
    </li>
  </ul>
  <ul
    class="cuke-pagination cuke-pagination-default"
    current="1"
  >
    <li
      class="cuke-pagination-show-total"
    />
    <li
      class="cuke-pagination-item cuke-pagination-prev cuke-pagination-item-disabled"
    >
      <svg
        fill="none"
        height="1em"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1em"
      >
        <polyline
          points="15 18 9 12 15 6"
        />
      </svg>
    </li>
    <li
      class="cuke-pagination-item cuke-pagination-item-selected"
    >
      1
    </li>
    <li
      class="cuke-pagination-item"
    >
      2
    </li>
    <li
      class="cuke-pagination-item"
    >
      3
    </li>
    <li
      class="cuke-pagination-item"
    >
      4
    </li>
    <li
      class="cuke-pagination-item"
    >
      5
    </li>
    <li
      class="cuke-pagination-item cuke-pagination-next"
    >
      <svg
        fill="none"
        height="1em"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1em"
      >
        <polyline
          points="9 18 15 12 9 6"
        />
      </svg>
    </li>
    <span>
      跳至
    </span>
    <input
      class="cuke-input cuke-number-input cuke-pagination-quick-jumper"
      placeholder=""
      type="text"
      value=""
    />
    <span>
      页
    </span>
  </ul>
  <ul
    class="cuke-pagination cuke-pagination-default"
    current="1"
  >
    <li
      class="cuke-pagination-show-total"
    />
    <li
      class="cuke-pagination-item cuke-pagination-prev cuke-pagination-item-disabled"
    >
      <svg
        fill="none"
        height="1em"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1em"
      >
        <polyline
          points="15 18 9 12 15 6"
        />
      </svg>
    </li>
    <li
      class="cuke-pagination-item cuke-pagination-item-selected"
    >
      1
    </li>
    <li
      class="cuke-pagination-item"
    >
      2
    </li>
    <li
      class="cuke-pagination-item"
    >
      3
    </li>
    <li
      class="cuke-pagination-item"
    >
      4
    </li>
    <li
      class="cuke-pagination-item"
    >
      5
    </li>
    <li
      class="cuke-pagination-item cuke-pagination-next"
    >
      <svg
        fill="none"
        height="1em"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1em"
      >
        <polyline
          points="9 18 15 12 9 6"
        />
      </svg>
    </li>
    <div
      class="cuke-select cuke-pagination-size-changer"
      position="bottom"
      value="[object Object]"
    >
      <div
        class="cuke-select-inner"
      >
        <div
          class="cuke-input-wrapper cuke-input-has-suffix"
        >
          <input
            class="cuke-input cuke-select-input"
            placeholder=""
            readonly=""
            type="text"
            value="10 条 / 页"
          />
          <span
            class="cuke-input-suffix"
          >
            <svg
              class="cuke-select-arrow"
              fill="none"
              height="1em"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="1em"
            >
              <polyline
                points="6 9 12 15 18 9"
              />
            </svg>
          </span>
        </div>
      </div>
      <div
        class="cuke-select-content cuke-pagination-size-changer-container cuke-pagination-size-changer-container-default cuke-select-close cuke-ui-no-animate"
        style="width:0;left:0;top:0"
      >
        <div
          class="cuke-select-option cuke-select-option-selected"
        >
          10 条 / 页
        </div>
        <div
          class="cuke-select-option"
        >
          20 条 / 页
        </div>
        <div
          class="cuke-select-option"
        >
          30 条 / 页
        </div>
        <div
          class="cuke-select-option"
        >
          40 条 / 页
        </div>
      </div>
    </div>
  </ul>
</div>
`;

exports[`<Pagination/> should render simple Pagination 1`] = `
<div
  class="cuke-pagination cuke-pagination-simple"
  current="1"
>
  <button
    class="cuke-button cuke-button-normal cuke-button-disabled cuke-button-size-small"
    disabled=""
    type="button"
  >
    <span>
      <svg
        fill="none"
        height="1em"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1em"
      >
        <polyline
          points="15 18 9 12 15 6"
        />
      </svg>
    </span>
  </button>
  <span
    class="cuke-pagination-simple-pages"
  >
    <span
      class="cuke-pagination-simple-page-index"
    >
      1
    </span>
    <span
      class="cuke-pagination-simple-page-separator"
    >
      |
    </span>
    <span>
      10
    </span>
  </span>
  <button
    class="cuke-button cuke-button-primary cuke-button-size-small"
    type="button"
  >
    <span>
      <svg
        fill="none"
        height="1em"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1em"
      >
        <polyline
          points="9 18 15 12 9 6"
        />
      </svg>
    </span>
  </button>
</div>
`;

exports[`<Pagination/> should render sizes 1`] = `
<div>
  <ul
    class="cuke-pagination cuke-pagination-small"
    current="1"
  >
    <li
      class="cuke-pagination-show-total"
    />
    <li
      class="cuke-pagination-item cuke-pagination-prev cuke-pagination-item-small cuke-pagination-item-disabled"
    >
      <svg
        fill="none"
        height="1em"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1em"
      >
        <polyline
          points="15 18 9 12 15 6"
        />
      </svg>
    </li>
    <li
      class="cuke-pagination-item cuke-pagination-item-selected cuke-pagination-item-small"
    >
      1
    </li>
    <li
      class="cuke-pagination-item cuke-pagination-item-small"
    >
      2
    </li>
    <li
      class="cuke-pagination-item cuke-pagination-item-small"
    >
      3
    </li>
    <li
      class="cuke-pagination-item cuke-pagination-item-small"
    >
      4
    </li>
    <li
      class="cuke-pagination-item cuke-pagination-item-small"
    >
      5
    </li>
    <li
      class="cuke-pagination-item cuke-pagination-next cuke-pagination-item-small"
    >
      <svg
        fill="none"
        height="1em"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1em"
      >
        <polyline
          points="9 18 15 12 9 6"
        />
      </svg>
    </li>
    <div
      class="cuke-select cuke-pagination-size-changer"
      position="bottom"
      value="[object Object]"
    >
      <div
        class="cuke-select-inner"
      >
        <div
          class="cuke-input-wrapper cuke-input-has-suffix"
        >
          <input
            class="cuke-input cuke-select-input cuke-input-small"
            placeholder=""
            readonly=""
            type="text"
            value="10 条 / 页"
          />
          <span
            class="cuke-input-suffix"
          >
            <svg
              class="cuke-select-arrow"
              fill="none"
              height="1em"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="1em"
            >
              <polyline
                points="6 9 12 15 18 9"
              />
            </svg>
          </span>
        </div>
      </div>
      <div
        class="cuke-select-content cuke-pagination-size-changer-container cuke-pagination-size-changer-container-small cuke-select-close cuke-ui-no-animate"
        style="width:0;left:0;top:0"
      >
        <div
          class="cuke-select-option cuke-select-option-selected"
        >
          10 条 / 页
        </div>
        <div
          class="cuke-select-option"
        >
          20 条 / 页
        </div>
        <div
          class="cuke-select-option"
        >
          30 条 / 页
        </div>
        <div
          class="cuke-select-option"
        >
          40 条 / 页
        </div>
      </div>
    </div>
  </ul>
  <ul
    class="cuke-pagination cuke-pagination-default"
    current="1"
  >
    <li
      class="cuke-pagination-show-total"
    />
    <li
      class="cuke-pagination-item cuke-pagination-prev cuke-pagination-item-disabled"
    >
      <svg
        fill="none"
        height="1em"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1em"
      >
        <polyline
          points="15 18 9 12 15 6"
        />
      </svg>
    </li>
    <li
      class="cuke-pagination-item cuke-pagination-item-selected"
    >
      1
    </li>
    <li
      class="cuke-pagination-item"
    >
      2
    </li>
    <li
      class="cuke-pagination-item"
    >
      3
    </li>
    <li
      class="cuke-pagination-item"
    >
      4
    </li>
    <li
      class="cuke-pagination-item"
    >
      5
    </li>
    <li
      class="cuke-pagination-item cuke-pagination-next"
    >
      <svg
        fill="none"
        height="1em"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1em"
      >
        <polyline
          points="9 18 15 12 9 6"
        />
      </svg>
    </li>
    <div
      class="cuke-select cuke-pagination-size-changer"
      position="bottom"
      value="[object Object]"
    >
      <div
        class="cuke-select-inner"
      >
        <div
          class="cuke-input-wrapper cuke-input-has-suffix"
        >
          <input
            class="cuke-input cuke-select-input"
            placeholder=""
            readonly=""
            type="text"
            value="10 条 / 页"
          />
          <span
            class="cuke-input-suffix"
          >
            <svg
              class="cuke-select-arrow"
              fill="none"
              height="1em"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="1em"
            >
              <polyline
                points="6 9 12 15 18 9"
              />
            </svg>
          </span>
        </div>
      </div>
      <div
        class="cuke-select-content cuke-pagination-size-changer-container cuke-pagination-size-changer-container-default cuke-select-close cuke-ui-no-animate"
        style="width:0;left:0;top:0"
      >
        <div
          class="cuke-select-option cuke-select-option-selected"
        >
          10 条 / 页
        </div>
        <div
          class="cuke-select-option"
        >
          20 条 / 页
        </div>
        <div
          class="cuke-select-option"
        >
          30 条 / 页
        </div>
        <div
          class="cuke-select-option"
        >
          40 条 / 页
        </div>
      </div>
    </div>
  </ul>
  <ul
    class="cuke-pagination cuke-pagination-large"
    current="1"
  >
    <li
      class="cuke-pagination-show-total"
    />
    <li
      class="cuke-pagination-item cuke-pagination-prev cuke-pagination-item-large cuke-pagination-item-disabled"
    >
      <svg
        fill="none"
        height="1em"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1em"
      >
        <polyline
          points="15 18 9 12 15 6"
        />
      </svg>
    </li>
    <li
      class="cuke-pagination-item cuke-pagination-item-selected cuke-pagination-item-large"
    >
      1
    </li>
    <li
      class="cuke-pagination-item cuke-pagination-item-large"
    >
      2
    </li>
    <li
      class="cuke-pagination-item cuke-pagination-item-large"
    >
      3
    </li>
    <li
      class="cuke-pagination-item cuke-pagination-item-large"
    >
      4
    </li>
    <li
      class="cuke-pagination-item cuke-pagination-item-large"
    >
      5
    </li>
    <li
      class="cuke-pagination-item cuke-pagination-next cuke-pagination-item-large"
    >
      <svg
        fill="none"
        height="1em"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1em"
      >
        <polyline
          points="9 18 15 12 9 6"
        />
      </svg>
    </li>
    <div
      class="cuke-select cuke-pagination-size-changer"
      position="bottom"
      value="[object Object]"
    >
      <div
        class="cuke-select-inner"
      >
        <div
          class="cuke-input-wrapper cuke-input-has-suffix"
        >
          <input
            class="cuke-input cuke-select-input cuke-input-large"
            placeholder=""
            readonly=""
            type="text"
            value="10 条 / 页"
          />
          <span
            class="cuke-input-suffix"
          >
            <svg
              class="cuke-select-arrow"
              fill="none"
              height="1em"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="1em"
            >
              <polyline
                points="6 9 12 15 18 9"
              />
            </svg>
          </span>
        </div>
      </div>
      <div
        class="cuke-select-content cuke-pagination-size-changer-container cuke-pagination-size-changer-container-large cuke-select-close cuke-ui-no-animate"
        style="width:0;left:0;top:0"
      >
        <div
          class="cuke-select-option cuke-select-option-selected"
        >
          10 条 / 页
        </div>
        <div
          class="cuke-select-option"
        >
          20 条 / 页
        </div>
        <div
          class="cuke-select-option"
        >
          30 条 / 页
        </div>
        <div
          class="cuke-select-option"
        >
          40 条 / 页
        </div>
      </div>
    </div>
  </ul>
</div>
`;
